### PR TITLE
[core] Introduce dynamic partition overwrite as default behavior

### PR DESCRIPTION
**Pull request by yuzelin**
### Purpose

Currently, `INSERT OVERWRITE` in Paimon will first purge all partitions specified in `PARTITION (...)`. In hive, the overwrite statement only purges partitions appeared in new data by default (called dynamic partition overwrite).
This PR introduce dynamic partition overwrite as default behavior.

### Tests
ReadWriteTableITCase:
- `testOverwriteNothing`
- `testSpecifySinglePartitionKey`
- `testDynamicOverwrite`

### API and Format 
CoreOptions:
- streaming-read-overwrite -> overwrite.support-streaming-read
- New: overwrite.dynamic-partition

Common:
`PredicateBuilder#partition(BinaryRow partition, RowType rowType)`

### Documentation

Modified: how-to#Writing Tables

### CODE REVIEW

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateBuilder.java [hunk 1]
```diff
@@ -344,4 +346,24 @@ public static Predicate partition(Map<String, String> map, RowType rowType) {
         }
         return predicate;
     }
+
+    public static Predicate partition(BinaryRow partition, RowType rowType) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/predicate/PredicateBuilder.java:350]: RowType partitionType

File: docs/content/how-to/writing-tables.md [hunk 2]
```diff
@@ -80,8 +80,10 @@ For more information, please check the syntax document:
 [Spark INSERT Statement](https://spark.apache.org/docs/latest/sql-ref-syntax-dml-insert-table.html)
 
 {{< hint info >}}
-Streaming reading will ignore the commits generated by `INSERT OVERWRITE` by default. If you want to read the
-commits of `OVERWRITE`, you can configure `streaming-read-overwrite`.
+1. Streaming reading will ignore the commits generated by `INSERT OVERWRITE` by default. If you want to read the
+commits of `OVERWRITE`, you can configure `overwrite.support-streaming-read`.
```
[the remark this answers is not in the document]
Comment (yuzelin) [reply docs/content/how-to/writing-tables.md:84]: No problem.

**Comment by yuzelin** [comment]
278C6CA1-099E-4ED8-974D-11F174C3E2BA

verified.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**